### PR TITLE
DVDVideoCodecAmlogic: add aspect ratio tracing for H.264

### PR DESCRIPTION
**Pull request by kszaq**
## Description
This set of patches add frame rate aspect tracking for H.264 videos. It is necessary for some DVB streams with variable aspect ratio.

## How Has This Been Tested?
LibreELEC community builds for Amlogic S805, S905 and S912 with Kodi v16, v17 and master.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/BitstreamConverter.cpp [hunk 1]
```diff
@@ -1224,6 +1236,316 @@ bool CBitstreamConverter::mpeg2_sequence_header(const uint8_t *data, const uint3
...
+    uint8_t aspect_ratio_idc = 0;
+    uint8_t separate_colour_plane_flag = 0;
+    int8_t frame_mbs_only_flag = -1;
+    unsigned int pic_width, pic_width_cropped;
+    unsigned int pic_height, pic_height_cropped;
+    unsigned int frame_crop_right_offset = 0;
+    unsigned int frame_crop_bottom_offset = 0;
+    unsigned int sar_width = 0;
+    unsigned int sar_height = 0;
+    uint32_t unitsInTick = 0;
```
Comment (Rechi) [inline xbmc/utils/BitstreamConverter.cpp]: unused but set variable

File: xbmc/utils/BitstreamConverter.cpp [hunk 2]
```diff
@@ -1224,6 +1236,316 @@ bool CBitstreamConverter::mpeg2_sequence_header(const uint8_t *data, const uint3
...
+    uint32_t timeScale = 0;
```
Comment (Rechi) [inline xbmc/utils/BitstreamConverter.cpp]: unused but set variable

**Comment by peak3d** [comment]
@kszaq do you have sample stream's (h.264) where these issues appear?

Beside this I don't understand the logic of the bitstream parser: assume a stream comes with initial aspect ratio set, then BitstreamParser is not scanning for aspect changes. It is only working if the initial stream comes with aspect=0.0, is this really wanted?

**Comment by kszaq** [comment]
@peak3d The parser is needed only for live TV where aspect ratio is not passed in hints. I will try to record a sample to expose the issue and ping back.

We shouldn't parse streams that have aspect ratio defined as the parser is a bit heavy on CPU and causes unwatchable 4K videos.

**Comment by kszaq** [comment]
PR updated: removed unused but set variables.

**Comment by MartijnKaijser** [comment]
@peak3d all good?

**Comment by peak3d** [comment]
@kszaq for me it feels wrong to make this check on demuxed package level.
It feels wrong because we don't hit the root cause.

Root cause is that the demuxer (is it pvr demux client)?? has to react to program change information in the ts stream and then reopen the stream (wich leads automatically to a new search of SPS inside ffmpeg)

The solution you provide is (halways / 4K) working only for AML, but all other platforms do still have the same issue.

Can you pls specify which addon / pvr you are focussing at here?
And pls. provide a small example ts stream snipped where we have aspect change during the stream.

Thanks!

**Comment by kszaq** [comment]
@peak3d This thread on LE forum is the only information I have for now: https://forum.libreelec.tv/thread/1542-solved-incorrect-aspect-ratio-with-some-tvheadend-channels/

I will see if I can reproduce the issue with any of DVB-T channel I can receive.

**Comment by peak3d** [comment]
@kszaq what pvr addon do you use for testing?
I have only HTS available but the streams provide here in germany normaly have no changes - at least I haven't recognized them so far....

**Comment by kszaq** [comment]
@peak3d I am using TVHeadend. 

Here is a short sample that should switch from 16:9 to 4:3: https://kszaq.libreelec.tv/samples/ar_from_169_to_43.ts

And it looks like I need to rethink the latest commit as that sample _does_ have `hints.aspect` but requires a switch.

**Comment by MrMC** [comment]
Could also just straight cmp of old vs new sps. if new sps does not match old. restart decoder. Can also two step the cmp, if length changes, it's different, if length is same and contents do not match, it's different.

This skips have to actually parse the sps which is the CPU intensive part. 

We do that in MrMC for VideoToolBox decoder and the above sample file switches from 16:9 to 4:3 as expected.

See https://github.com/MrMC/mrmc/blob/master/xbmc/cores/dvdplayer/DVDCodecs/Video/DVDVideoCodecVideoToolBox.cpp

ValidateVTSessionParameterSetsForRestart

**Comment by kszaq** [comment]
@MrMC Sounds like a good idea but unfortunately it's over my skills to implement. 

@peak3d What do you think? Perhaps this PR should be dropped and a better solution is possible?

**Comment by peak3d** [comment]
@kszaq 
1.) the video starts with ar=0.0 on my vanilla kodi master, but displayed wrong.
2.) I'm currently investigating this issue in regard of scanning for SPS only if a ts program header was send before.
3.) parsing SPS is normally the job of the decoder, so question is if AML provides AR information maybe already?

**Comment by kszaq** [comment]
3) I'm not sure, I'll try to figure that one out.

**Pull request closed without merging**